### PR TITLE
boards: esp32: Add support for M5Stack AtomS3

### PR DESCRIPTION
**Pull request by kartben**
This introduces support for the ESP32S3 based M5Stack AtomS3.

Tested all of the following:

- [x] LCD display and accelerometer tested with `samples/modules/lvgl/accelerometer_chart`, which works out-of-the-box (see below!)
- [x] Button (`samples/basic/button` and `samples/subsys/display/lvgl` sample)
- [x] Bluetooth (`samples/bluetooth/peripheral_hr`)
- [x] Wi-Fi (`samples/net/wifi`, wifi scan works)
- [x] Grove connector, tested with `samples/sensor/sensor_shell` and an I2C device attached

Doc page: https://builds.zephyrproject.io/zephyr/pr/64434/docs/boards/xtensa/m5stack_atoms3/doc/index.html

[Pxl 20231026 101330936.webm](https://github.com/zephyrproject-rtos/zephyr/assets/128251/d233b3cb-3746-4376-accc-271346083388)

### CODE REVIEW

**Comment by kartben** [comment]
The first two commits are part of PR #64451 so you may ignore them.

**Comment by sylvioalves** [comment]
@kartben awesome!

**Review by sylvioalves — DISMISSED** [review]

**Comment by kartben** [comment]
Rebased as the 2 commits mentioned earlier are not needed anymore now that 1fac5ed2a60b5365953e74c6dbc1ba3640b728ad fixed the issue with Xtensa CI.

**Review by sylvioalves — DISMISSED** [review]

**Comment by kartben** [comment]
@sylvioalves should I exclude this board from tests/lib/heap testcase, like other (but not all??) esp32-s2 and s3? I don't think the test failed in earlier runs though, but not 100% sure.

**Comment by kartben** [comment]
@sylvioalves @marekmatej what should I be doing with the failing heap test? should I just add this board as a `platform_exclude` in the testcase? There are other ESP32S3 that are working just fine (but also some that are excluded).

```
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr/zephyr_pre0.map 
: && ccache /opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/lib/posix/liblib__posix.a  zephyr/subsys/random/libsubsys__random.a  zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/regulator/libdrivers__regulator.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0"  -L/__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -lgcc  -Wl,--fatal-warnings  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_INTEGER_PRINTF_SCANF  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/src/linker/esp32s3.rom.alias.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/soc/esp32s3/ld/esp32s3.peripherals.ld  -Wl,--wrap=k_calloc  -Wl,--wrap=k_malloc  --specs=picolibc.specs  -lc  -lgcc && cd /__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr && /usr/bin/cmake -E true
/opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `.dram0.data' will not fit in region `dram0_0_seg'
/opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `dram0_0_seg' overflowed by 1536 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

File: boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+CONFIG_GPIO=y
+CONFIG_REGULATOR=y # for LCD backlight
+
+CONFIG_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_UART_CONSOLE=y
+
+# for debugging
+CONFIG_SHELL=y
```
Comment (sylvioalves) [inline boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig:18]: I suggest removing this as default config. This will allow `test/lib/heap` to pass. User can manually enable this if necessary. Sounds good?

File: boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_BOARD_M5STACK_ATOMS3=y
+CONFIG_SOC_SERIES_ESP32S3=y
+
+CONFIG_MAIN_STACK_SIZE=2048
+
+CONFIG_ESP_HEAP_MEM_POOL_REGION_1_SIZE=0
```
Comment (sylvioalves) [inline boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig:8]: This is not needed nor available for ESP32S3. Only for ESP32 that has dual ram bank. So you can remove it from here.

**Comment by sylvioalves** [comment]
> @sylvioalves should I exclude this board from tests/lib/heap testcase, like other (but not all??) esp32-s2 and s3? I don't think the test failed in earlier runs though, but not 100% sure.

If you remove the `CONGI_SHELL=y`, test will pass.

File: boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig (hunk 1 above)
Comment (kartben) [reply boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig:18]: Sure thing! Definitely the right thing to do. Pretty sure I copy pasted from another esp32 board fwiw, including the "for debugging" comment.

**Comment by kartben** [comment]
> > @sylvioalves should I exclude this board from tests/lib/heap testcase, like other (but not all??) esp32-s2 and s3? I don't think the test failed in earlier runs though, but not 100% sure.
> 
> If you remove the `CONFIG_SHELL=y`, test will pass.

Thanks for your help! All green now!
Let me know who else you think could help reviewing this PR btw, as it hasn't had a lot of traction in the past two weeks. Thanks!

**Review by sylvioalves — DISMISSED** [review]

**Comment by sylvioalves** [comment]
@marekmatej @LucasTambor PTAL

File: boards/xtensa/m5stack_atoms3/support/openocd.cfg
Comment (LucasTambor) [inline boards/xtensa/m5stack_atoms3/support/openocd.cfg:1]: Is this file necessary? In the index.rst says:

```
M5Stack AtomS3 debugging is not supported due to pinout limitations.
```

Comment (kartben) [reply boards/xtensa/m5stack_atoms3/support/openocd.cfg:1]: Ha! Good catch. Will remove.

File: boards/xtensa/m5stack_atoms3/m5stack_atoms3.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  - regulator
+  - uart
+  - pinmux
+  - nvs
+  - display
+testing:
+  ignore_tags:
+    - net
+    - bluetooth
+vendor: espressif
```
Comment (sylvioalves) [inline boards/xtensa/m5stack_atoms3/m5stack_atoms3.yaml:21]: @kartben, nitpick by @LucasTambor : vendor here should be `m5stack` instead.

Comment (kartben) [reply boards/xtensa/m5stack_atoms3/m5stack_atoms3.yaml:21]: not a nitpick at all - thanks! I will update all other m5stack boards accordingly.

File: boards/xtensa/m5stack_atoms3/support/openocd.cfg
Comment (kartben) [reply boards/xtensa/m5stack_atoms3/support/openocd.cfg:1]: done, thx!

**Review by LucasTambor — APPROVED** [review]

File: boards/xtensa/m5stack_atoms3/board.cmake [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(NOT "${OPENOCD}" MATCHES "^${ESPRESSIF_TOOLCHAIN_PATH}/.*")
+  set(OPENOCD OPENOCD-NOTFOUND)
+endif()
+find_program(OPENOCD openocd PATHS ${ESPRESSIF_TOOLCHAIN_PATH}/openocd-esp32/bin NO_DEFAULT_PATH)
```
Comment (marekmatej) [inline boards/xtensa/m5stack_atoms3/board.cmake:6]: Since debugging is not supported, maybe consider reducing this file to single line:
`include(${ZEPHYR_BASE}/boards/common/esp32.board.cmake)`

File: boards/xtensa/m5stack_atoms3/m5stack_atoms3-pinctrl.dtsi [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	spim2_default: spim2_default {
+		group1 {
+			pinmux = <SPIM2_SCLK_GPIO17>,
+				 <SPIM2_CSEL_GPIO15>;
+		};
+		group2 {
+			pinmux = <SPIM2_MOSI_GPIO21>;
+			output-low;
+		};
+
```
Comment (marekmatej) [inline boards/xtensa/m5stack_atoms3/m5stack_atoms3-pinctrl.dtsi:52]: remove blank line

**Review by marekmatej — COMMENTED** [review]
some nitpicks, but looks good

**Review by marekmatej — APPROVED** [review]
